addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/NeeoDeviceProtocol.java

```diff
@@ -0,0 +1,125 @@
+
+    /** The {@link NeeoRoom} */
+    private final NeeoRoom neeoRoom;
+
+    /** The {@link NeeoDevice} */
+    private final NeeoDevice neeoDevice;
+
+    /**
+     * Instantiates a new neeo device protocol.
+     *
+     * @param callback the non-null callback
+     * @param roomKey the non-empty room key
+     * @param deviceKey the non-empty device key
+     * @throws IOException Signals that an I/O exception has occurred.
+     */
+    public NeeoDeviceProtocol(NeeoHandlerCallback callback, String roomKey, String deviceKey) throws IOException {
+        Objects.requireNonNull(callback, "callback cannot be null");
+        Objects.requireNonNull(roomKey, "roomKey cannot be empty");
+        Objects.requireNonNull(deviceKey, "deviceKey cannot be empty");
+
+        this.callback = callback;
+
+        final NeeoBrainApi api = callback.getApi();
+        if (api == null) {
+            throw new IllegalArgumentException("NeeoBrainApi cannot be null");
+        }
+
+        neeoRoom = api.getRoom(roomKey);
+        if (neeoRoom == null) {
+            throw new IllegalArgumentException("Room (" + roomKey + ") was not found in the NEEO Brain");
+        }
+
+        neeoDevice = neeoRoom.getDevices().getDevice(deviceKey);
+        if (neeoDevice == null) {
+            throw new IllegalArgumentException(
+                    "Device (" + deviceKey + ") was not found in the NEEO Brain for room (" + roomKey + ")");
+        }
+    }
+
+    /**
+     * Returns the callback being used
+     *
+     * @return the non-null callback
+     */
+    public NeeoHandlerCallback getCallback() {
+        return callback;
+    }
+
+    /**
+     * Refresh the macro status.
+     *
+     * @param macroKey the non-null macro key
+     */
+    public void refreshMacroStatus(String macroKey) {
+        NeeoUtil.requireNotEmpty(macroKey, "macroKey cannot be empty");
+
+        final NeeoMacro macro = neeoDevice.getMacros().getMacro(macroKey);
+        if (macro != null) {
+            callback.stateChanged(UidUtils.createChannelId(NeeoConstants.DEVICE_GROUP_MACROSID, macroKey),
+                    OnOffType.OFF);
+        }
+    }
+
+    /**
+     * Sets the macro status. If the status is true, the macro will be triggered. If false, nothing occurs
+     *
+     * @param macroKey the non-null macro key
+     * @param start whether to start (true) or stop (false) the recipe
+     */
+    public void setMacroStatus(String macroKey, boolean start) {
+        NeeoUtil.requireNotEmpty(macroKey, "macroKey cannot be empty");
+
+        final NeeoBrainApi api = callback.getApi();
+        if (api == null) {
+            logger.debug("API is null [likely bridge is offline]");
+        } else {
+            try {
+                if (start) {
+                    api.triggerMacro(neeoRoom.getKey(), neeoDevice.getKey(), macroKey);
+                    callback.scheduleTask(() -> {
+                        refreshMacroStatus(macroKey);
+                    }, 500);
```

Comment:
What if the device takes longer than 500ms to change its status based on the sent command? Will you miss that the device has accepted/not accepted the desired state change?

If the brain doesn't provide any push mechanism, wouldn't it make sense to have a polling job running that polls the status every x seconds and set the polled state in OH2? The hue binding does that for example, because the bridge also doesn't have a push mechanism.